### PR TITLE
Pipes - Eva - api-muncher

### PR DESCRIPTION
**Pull request by Esawyer25**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| How did you go about exploring the Edamam API, how did you try querying the API? |I used postman some, but also just used HTTParty calls from terminal sometimes since it seemed like sometimes i would get slightly different responses.  |
| Describe your API Wrapper.  How did you decide on the methods you created? | I created a method for every time I needed to call the API, which was once to get all the results, and once to get a specific recipe from the list of all results.  |
| Describe an edge case or failure case test you wrote for your API Wrapper. | I wrote a test for if the token was bogus.  I tried to write a test for if the query produced no recipes, but that functionality kept interfering with my test for if the token was bogus and I did not have time make those two different cases lead to two different results. |
| Explain how VCR aids in testing an API. | They record responses from set queries to the API so that we can use the recorded responses instead of re-asking the API every time we test.  |
| What is the Heroku URL of your deployed application? |  https://floating-citadel-58194.herokuapp.com/ |

### CODE REVIEW

File: lib/edemam_api_wrapper.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    image = api_params["image"]
+    source = api_params["source"]
+    uri = api_params["uri"]
+    return Recipe.new(label, recipe_url, ingredients, dietary, image, source, uri)
+  end
+
+  def self.check_status(hits)
+    if hits == []
+      raise ApiError.new("API call to slack failed")
+    end
```
Comment (droberts-sea) [inline lib/edemam_api_wrapper.rb:52]: Here you throw an error, but in your controller you never catch it. This means that the exception makes it all the way to the user, instead of a polite and actionable error message.

File: lib/edemam_api_wrapper.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
+#require "httparty"
+#require 'pry'
+
+class EdemamApiWrapper
+  BASE_URL = "https://api.edamam.com/search?"
+  APP_ID = ENV["APP_ID"]
+  APP_KEY= ENV["APP_KEY"]
+  END_URL = "&app_id=" + APP_ID + "&app_key=" + APP_KEY
+  URI_BASE ="http://www.edamam.com/ontologies/edamam.owl%23recipe_"
```
Comment (droberts-sea) [inline lib/edemam_api_wrapper.rb:9]: Having two variables named `URI_BASE` and `BASE_URL` seems confusing to me. I would probably have renamed `URI_BASE` to something like `RECIPE_URI_PREFIX` to make them more visually distinct.

File: test/controllers/recipe_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      end
+    end
+
+    # it "does something when there is no recipe" do
+    #   VCR.use_cassette("recipes") do
+    #     uri ="_bogus"
+    #     get recipe_path(uri)
+    #   end
+    # end
+    #I choose not to do this b/c I don't see how this could happen unless someone was typing bad info right into the address bar.
```
Comment (droberts-sea) [inline test/controllers/recipe_controller_test.rb:21]: Even if the answer is "produce an error", I still think "what does my app do when a user manually mangles the URL" is an interesting question.

One way this could occur in the wild: Edamam changes the format of their URIs, but a user has a bookmark using the old style.

File: test/controllers/recipe_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  end
+
+
+  describe "index" do
+    it "returns a list of recipes if there are any related to the search" do
+      VCR.use_cassette("recipes") do
+        params = {}
+        params[:query] = "chicken"
```
Comment (droberts-sea) [inline test/controllers/recipe_controller_test.rb:29]: This is a good success case, but there are a few more interesting pieces you're not testing:

1. What if the search returns no results (for example, a search for "dfaljdfalflaksjflaksj")?
1. Does the page still load if you pass paging parameters?
1. What if the paging parameters are invalid (negative number, etc)?
1. What if the paging parameters take you past the end of the available search results?

File: test/lib/edemam_api_wrapper_test.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        uri ="http://www.edamam.com/ontologies/edamam.owl%23recipe_7bf4a371c6884d809682a72808da7dc2"
+        result = EdemamApiWrapper.find_a_recipe(uri)
+        result.must_be_kind_of Recipe
+        result.label.must_equal "Teriyaki Chicken"
+      end
+    end
+
+    it "Raises an ApiError when the uri is bogus" do
+      VCR.use_cassette("recipes") do
+        uri ="http://www.edamam.com/ontologies/edamam.owl%23recipe_7bf4a371c6884d809682abogus"
```
Comment (droberts-sea) [inline test/lib/edemam_api_wrapper_test.rb:37]: Good error case

File: test/lib/edemam_api_wrapper_test.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+      VCR.use_cassette("recipes") do
+        search_term = "chicken"
+        response = EdemamApiWrapper.list_recipes(search_term)
+        response.count.must_be :>, 0
+      end
+    end
+
+    # it "Raises an ApiError when the token is bad" do
+    #   VCR.use_cassette("recipes") do
+    #      proc {
```
Comment (droberts-sea) [inline test/lib/edemam_api_wrapper_test.rb:16]: I like the idea of this test. Probably the easiest way to make it work would be to use a different cassette.

File: test/lib/edemam_api_wrapper_test.rb [hunk 7]
```diff
@@ -0,0 +1,45 @@
+require 'test_helper'
+
+describe EdemamApiWrapper do
+
+  describe "list_recipes" do
+    it "Can find recipes" do
+      VCR.use_cassette("recipes") do
+        search_term = "chicken"
```
Comment (droberts-sea) [inline test/lib/edemam_api_wrapper_test.rb:8]: Missing cases here are similar to the ones for your `index` controller action.

**Comment by droberts-sea** [comment]
# API Muncher
## What We're Looking For

| Feature | Feedback    |
| :------------- | :------------- |
| **Core Requirements** |   |
| Git hygiene | yes |
| Comprehension questions	| yes |
| **General** |  |
| Rails fundamentals (RESTful routing, use of named paths) | yes |
| Semantic HTML | yes |
| Errors are reported to the user | no - It looks like you've got the `flash` infrastructure in place, but if enter a search with no results or mangle a recipe URI I get an exception. |
| API Wrapper to handle the API requests | yes |
| Controller testing | some - see inline comments |
| Lib testing | yes |
| Search Functionality | yes |
| List Functionality | yes |
| Show individual item functionality (link to original recipe opens in new tab) | yes |
| **Styling** |  |
| Foundation Styling for responsive layout | yes |
| List View shows 10 items at a time/pagination | yes |
| The app is styled to create an attractive user interface | yes |
| **API Features** | |
| The App attributes Edaman | yes |
| The VCR casettes do not contain the API key | yes |
| **External Resources** |  |
| Link to deployed app on Heroku | yes |
|  **Overall** |  |
